From dbe74dc5cb4a4d8be2732c3ac32ed6b5167991c5 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Tue, 26 Nov 2013 14:30:59 +0000
Subject: [PATCH 055/351] ASoC: wm_adsp: Ensure we don't attempt to create
 controls twice

wm_adsp_create_grouped_control is called from a loop which iterates
through the alg_regions list. However it also adds new regions into this
list, controls for these regions will be added twice, once when the new
region is added to the list and once when the iteration reached the new
region in the list. This patch changes this such that the control is
only added once it is processed from the alg_regions list.

Change-Id: If8d30f5dc1ce1b496c198e31cf7e3b350c646d25
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/wm_adsp.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/sound/soc/codecs/wm_adsp.c b/sound/soc/codecs/wm_adsp.c
index f31ceab..fc7c077 100644
--- a/sound/soc/codecs/wm_adsp.c
+++ b/sound/soc/codecs/wm_adsp.c
@@ -1102,7 +1102,6 @@ static int wm_adsp_create_grouped_control(struct wm_adsp *dsp,
 	struct wm_adsp_alg_region *r;
 	int ret;
 
-	region->block = 0;
 	/* This is the quick case for control groups of a single block */
 	if (region->len <= 512)
 		return wm_adsp_create_control(dsp, region);
@@ -1130,10 +1129,11 @@ static int wm_adsp_create_grouped_control(struct wm_adsp *dsp,
 		else
 			r->len = len - offset;
 		offset += r->len;
+		/* We do not need to create the control for this new
+		 * region as it will be created once we reach the region
+		 * whilst processing the alg_regions list.
+		 */
 		list_add_tail(&r->list, &dsp->alg_regions);
-		ret = wm_adsp_create_control(dsp, r);
-		if (ret < 0)
-			return ret;
 	} while (offset < len);
 
 	return 0;
-- 
1.7.9.5

